File: java/tree/UnionFind.java
import java.util.Map;
import java.util.HashMap;

public class UnionFind {
    
    Map<Integer, Integer> par;
    Map<Integer, Integer> rank;

    public UnionFind(int n) {
        par = new HashMap<>();
        rank = new HashMap<>();

        for (int i = 1; i < n + 1; i++) {
            par.put(i,i);
            rank.put(i,0);
        }
    }

    // Find parent of n, with path compression.
    public int find(int n) {
        int p = par.get(n);
        while (p != par.get(p)) {
            par.put(p, par.get(par.get(p)));
            p = par.get(p);
        }
        return p;
    }

    // Union by height / rank.
    // Return false if already connected, true otherwise.
    public boolean union(int n1, int n2) {
        int p1 = this.find(n1), p2 = this.find(n2);
        if (p1 == p2) {
            return false;
        }

        if (rank.get(p1) > rank.get(p2)) {
            par.put(p2, p1);
        } else if (rank.get(p1) < rank.get(p2)) {
            par.put(p1, p2);
        } else {
            par.put(p1, p2);
            rank.put(p2, rank.get(p2) + 1);
        }
        return true;
    }
}

File: swift/tree/unionFind.swift
class UnionFind {
  var par = [Int: Int]()
  var rank = [Int: Int]()

  init(_ n: Int) {
    for i in 1..<n + 1 {
      par[i] = i
      rank[i] = 0
    }
  }

  // Find parent of n, with path compression.
  func find(_ n: Int) -> Int {
    var p = par[n]!
    while p != par[p] {
      par[p] = par[par[p]!]
      p = par[p]!
    }
    return p
  }

  // Union by height / rank.
  // Return false if already connected, true otherwise.
  func union(_ n1: Int, _ n2: Int) -> Bool {
    let p1 = find(n1), p2 = find(n2)
    if p1 == p2 {
      return false
    }

    if rank[p1]! > rank[p2]! {
      par[p2] = p1
    }
    else if rank[p1]! < rank[p2]! {
      par[p1] = p2
    }
    else {
      par[p1] = p2
      rank[p2]! += 1
    }
    return true
  }
}